### PR TITLE
Logstash output

### PR DESCRIPTION
**Pull request by lykkin**
## What is the problem this PR solves?

Issue defined in https://github.com/elastic/fleet-server/issues/1185.

Fleet server currently assumes its default output will have an ES type and uses that information to update the output's API key. This PR separates the output preparation (e.g. API key updating) into its own abstraction that makes handling multiple outputs/different types easier.

## How does this PR solve the problem?

Encapsulates the policy output preparation, which discerns its actions based on the type of the output.

## How to test this PR locally

More tests to come soon.

## Checklist

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues
- Closes https://github.com/elastic/fleet-server/issues/1185

### CODE REVIEW

**Comment by defensivedepth** [comment]
Hello there! 

Will this functionality be available for 7.x?

**Comment by lykkin** [comment]
@defensivedepth Heya, thanks for the interest!

It's not currently scoped to be ported back to 7.x to my knowledge, though @nimarezainia might know more about the more distant horizon plans for backporting.

**Comment by defensivedepth** [comment]
thanks, yes, it would be very helpful to know if the longer-term plan is to backport this to 7.

Thanks!

**Comment by nimarezainia** [comment]
> thanks, yes, it would be very helpful to know if the longer-term plan is to backport this to 7.
> 
> Thanks!

No current plans to backport. Please let us know why you would need this in 7.x - i am currently not sure of dependancies.

**Comment by defensivedepth** [comment]
> > thanks, yes, it would be very helpful to know if the longer-term plan is to backport this to 7.
> > Thanks!
> 
> No current plans to backport. Please let us know why you would need this in 7.x - i am currently not sure of dependancies.

Just needed this information for planning, thanks!

File: internal/pkg/policy/policy_output_test.go [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+		Type: OutputTypeLogstash,
+		Name: "test output",
+		Role: &RoleT{
+			Sha2: "fake sha",
+			Raw:  TestPayload,
+		},
+	}
+
+	err := po.Prepare(context.Background(), zerolog.Logger{}, bulker, &model.Agent{}, smap.Map{}, false)
+	if err != nil {
```
Comment (ph) [inline internal/pkg/policy/policy_output_test.go:35]: I think we have the required module in fleet-server to do `require.NoNit(t, err, "expect....")`, also what is expected here, we check we aren't returning an error but we do not check for any transformation result?

File: internal/pkg/policy/policy_output_test.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+		Name: "test output",
+		Role: nil,
+	}
+
+	err := po.Prepare(context.Background(), zerolog.Logger{}, bulker, &model.Agent{}, smap.Map{}, false)
+	// No permissions are required by logstash currently
+	if err != nil {
+		t.Error("expected prepare to pass")
+	}
+}
```
Comment (ph) [inline internal/pkg/policy/policy_output_test.go:55]: What is the expected result from this test?

File: internal/pkg/policy/policy_output.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+					return err
+				}
+				agent.DefaultApiKey = outputAPIKey.Agent()
+			}
+		}
+	case OutputTypeLogstash:
+		zlog.Debug().Msg("preparing logstash output")
+		zlog.Info().Msg("no actions required for logstash output preparation")
+	default:
+		zlog.Debug().Msgf("unknown output type: %s; skipping preparation", p.Type)
```
Comment (ph) [inline internal/pkg/policy/policy_output.go:121]: Not sure what was the previous behavior, but this looks like we should _return an error_ and log it as an error. I supposed the only way to get into that section of the code is that the format received from Fleet is incorrect.

File: internal/pkg/policy/policy_output.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+			zlog.Error().Str("name", p.Name).Msg("policy does not contain required output permission section")
+			return ErrNoOutputPerms
+		}
+
+		// Determine whether we need to generate an output ApiKey.
+		// This is accomplished by comparing the sha2 hash stored in the agent
+		// record with the precalculated sha2 hash of the role.
+
+		// Note: This will need to be updated when doing multi-cluster elasticsearch support
+		// There will need to be multiple api keys tracked on the agent model
```
Comment (ph) [inline internal/pkg/policy/policy_output.go:56]: This sentence sounded unclear to me.

File: internal/pkg/policy/policy_output.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+				}
+				if agent.DefaultApiKeyId != "" {
+					fields[dl.FieldDefaultApiKeyHistory] = model.DefaultApiKeyHistoryItems{
+						Id:        agent.DefaultApiKeyId,
+						RetiredAt: time.Now().UTC().Format(time.RFC3339),
+					}
+				}
+				body, err := renderUpdatePainlessScript(fields)
+
+				// Using painless script to append the old keys to the history
```
Comment (ph) [inline internal/pkg/policy/policy_output.go:104]: We might want to move this before line 102?

File: internal/pkg/policy/parsed_policy_test.go [hunk 6]
```diff
@@ -514,13 +506,462 @@ func TestNewParsedPolicy(t *testing.T) {
...
 		expectedSha2 := "d4d0840fe28ca4900129a749b56cee729562c0a88c935192c659252b5b0d762a"
-		if pp.Default.Role.Sha2 != expectedSha2 {
-			t.Fatal(fmt.Sprintf("Expected sha2: '%s', got '%s'.", expectedSha2, pp.Default.Role.Sha2))
+		if defaultOutput.Role.Sha2 != expectedSha2 {
+			t.Fatal(fmt.Sprintf("Expected sha2: '%s', got '%s'.", expectedSha2, defaultOutput.Role.Sha2))
 		}
 	}
 }
+
+const noESOutputPolicy = `
```
Comment (ph) [inline internal/pkg/policy/parsed_policy_test.go:521]: Can we move that into an external file, This would make json linting and updating it easier?

**Review by ph — CHANGES_REQUESTED** [review]
I've did an initial pass in this PR.

Thanks for fixing some of the linting issues with the capitalization. Since you move code between files it makes it harder for the reviewer to see what logic has changed. From what I see I think this is correct. 

I've looked at the test for logstash (on the prepare operation) we seems to only check for error, should also either look at what should have changed or make sure calls are not executed.

One question concerning the behavior of fleet, when Logstash is used what name would be used in the output block? default, logstash or something else? I am talking about this section of the YAML:


```yaml
output:
  myls: # this part.
    type: logstash
```

File: internal/pkg/policy/policy_output_test.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+		t.Error("expected prepare to pass", err)
+	}
+
+	updatedKey, ok := policyMap.GetMap("test output")["api_key"].(*bulk.ApiKey)
+
+	if !ok {
+		t.Errorf("unable to cast api key")
+	}
+
+	if updatedKey.Key != bulker.MockedAPIKey.Key {
```
[the remark this answers is not in the document]
Comment (lykkin) [reply internal/pkg/policy/policy_output_test.go:164]: ah, it's in `require`, i thought it was a bit weird that there were no assertions :sweat_smile:

File: internal/pkg/policy/policy_output_test.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+		"test output": map[string]interface{}{},
+	}
+	testAgent := &model.Agent{}
+	err := po.Prepare(context.Background(), zerolog.Logger{}, bulker, testAgent, policyMap, true)
+	if err != nil {
+		t.Error("expected prepare to pass", err)
+	}
+
+	updatedKey, ok := policyMap.GetMap("test output")["api_key"].(*bulk.ApiKey)
+
```
[the remark this answers is not in the document]
Comment (lykkin) [reply internal/pkg/policy/policy_output_test.go:159]: ignoring causes linting errors, so might as well

File: internal/pkg/policy/policy_output_test.go (hunk 2 above)
Comment (lykkin) [reply internal/pkg/policy/policy_output_test.go:55]: similar to the one above, it's just a smoke test for the logstash output with no permissions.

all logstash tests will be smoke tests, as the prepare step doesn't do anything tangible.

**Comment by lykkin** [comment]
> I've did an initial pass in this PR.
> 
> Thanks for fixing some of the linting issues with the capitalization. Since you move code between files it makes it harder for the reviewer to see what logic has changed. From what I see I think this is correct.
> 
> I've looked at the test for logstash (on the prepare operation) we seems to only check for error, should also either look at what should have changed or make sure calls are not executed.
> 
> One question concerning the behavior of fleet, when Logstash is used what name would be used in the output block? default, logstash or something else? I am talking about this section of the YAML:
> 
> ```yaml
> output:
>   myls: # this part.
>     type: logstash
> ```

Yeah, currently all the logstash output tests are simple smoke tests, I could go either way on adding more expectations since they are no ops. I generally dislike looking for negative signals (i.e. something DIDN'T happen) unless there are configurations where they are expected to happen (e.g. update being called in the ES case).

For the naming, it will be `default` for now, but only because that's what the UI supports. There are no restrictions incurred by this code on the name of the output.

**Review by ph — CHANGES_REQUESTED** [review]
LGTM, only missing a changelog!

**Review by ph — APPROVED** [review]

**Pull request merged**